### PR TITLE
Issue #2749287: Implement a Calculator class and use it in the Price value object

### PR DESCRIPTION
**Pull request by FrancisBaileyH**

### CODE REVIEW

**Comment by bojanz** [comment]
Thanks! Do you mind filling out the docblocks? We need a description for each method, param, return value.

**Comment by FrancisBaileyH** [comment]
@bojanz I've gone ahead and filled out the doc blocks.

**Comment by FrancisBaileyH** [comment]
Fixed two docblocks with improper spacing.

File: modules/price/src/CurrencyMismatchException.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace Drupal\commerce_price;
+
+/**
+ * This exception is thrown when attempting to perform math using two different currencies.
+ * For example, adding a price in USD to another in EUR.
+ */
+class CurrencyMismatchException extends \RuntimeException {
```
Comment (mglaman) [inline modules/price/src/CurrencyMismatchException.php]: @bojanz should we have an Exception namespace for these?

Comment (FrancisBaileyH) [inline modules/price/src/CurrencyMismatchException.php]: That's an easy change so let me know if this is something you'd like implemented.

**Comment by FrancisBaileyH** [comment]
@bojanz @mglaman any updates on this pr?

File: modules/price/src/Price.php [hunk 2]
```diff
@@ -63,4 +64,214 @@ public function __toString() {
...
 
+  /**
+   * Convert a \Drupal\commerce_price\Price object to a different currency.
+   *
+   * @param string $currency
+   *   A currency code.
+   * @param string $rate
+   *   A currency rate corresponding to the currency code.
+   *
+   * @return Price
```
Comment (bojanz) [inline modules/price/src/Price.php]: Any object name given as a return or param hint must be fully qualified (have the full namespace).

File: modules/price/src/Price.php [hunk 3]
```diff
@@ -63,4 +64,214 @@ public function __toString() {
...
+   *   A new \Drupal\commerce_price\Price instance.
```
Comment (bojanz) [inline modules/price/src/Price.php]: You don't need to use the full namespace in the descriptions.

File: modules/price/src/CurrencyMismatchException.php (hunk 1 above)
Comment (bojanz) [inline modules/price/src/CurrencyMismatchException.php]: Yes, we should. Follows the example of other Exception classes in Commerce.

File: modules/price/src/InvalidArgumentException.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace Drupal\commerce_price;
+
+class InvalidArgumentException extends \InvalidArgumentException {
```
Comment (bojanz) [inline modules/price/src/InvalidArgumentException.php]: We can remove this one. It doesn't add anything over the builtin \InvalidArgumentException.

**Comment by FrancisBaileyH** [comment]
@bojanz okay I applied your suggested changes.

**Comment by bojanz** [comment]
Committed in 071d9ba6ce028d4e3a260082e997ded81d842e47 after some bike shedding :)

**Pull request closed without merging**